File: Data Structures & Algorithms/house-robber-ii/submission-4.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        
        return max(self.dfs(nums[:-1]), self.dfs(nums[1:]))
    
    def dfs(self, nums):
        p1, p2 = 0, 0

        for n in nums:
            temp = max(p1 + n, p2)
            p1 = p2
            p2 = temp
        return p2
    
# [2,9,8,3,6]

File: Data Structures & Algorithms/longest-consecutive-sequence/submission-4.java
class Solution {
    public int longestConsecutive(int[] nums) {
        Set<Integer> set = new HashSet<>();
        for(int n : nums) {
            set.add(n);
        }

        int res = 0;
        for(int i=0; i<nums.length; ++i) {
            int count = 0;
            if(!set.contains(nums[i] - 1)) {
                ++count;
                int cur = nums[i] + 1;
                while(set.contains(cur)) {
                    ++count;
                    ++cur;
                }
                res = Math.max(res, count);
            }
        }
        return res;
    }
}
